## Question1-Answer

In [47]:
A = [1, 2, 3, 4, 5, 6] # creating list A
B = [13, 21, 34] # creating list B

In [48]:
A_B = [1, 2, 3, 4, 5, 6, 13, 21, 34] #Creating list A_B

In [49]:
A.extend(B) # Extending list A with list B

In [50]:
A #show output of A after extending

[1, 2, 3, 4, 5, 6, 13, 21, 34]

In [51]:
assert A_B == A #  Checking if it's true that A.extend(B) is the same list as A_B

In [52]:
del A[6 : ] # removing extended value of A

In [53]:
A 

[1, 2, 3, 4, 5, 6]

In [54]:
A.append(B) #appending list B to A

In [55]:
A

[1, 2, 3, 4, 5, 6, [13, 21, 34]]

In [56]:
assert A_B == A #  Checking if it's true that A.extend(B) is the same list as A_B

AssertionError: 

__Therefore A_B = [1, 2, 3, 4, 5, 6] can be gotten by the option of A.extend(B)__

## Question2-Answer

In [11]:
import numpy as np

In [12]:
f = np.identity(3)

In [13]:
g = np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]])

In [14]:
print(f)
print(g)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
[[1 0 0]
 [0 1 0]
 [0 0 1]]


**Therefore np.identity(3) and np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]], dtype = float) can be used to create an identity
matrix in python.**

_Note: np.array([1, 0, 0], [0, 1, 0], [0, 0, 1]) will give a value error beacause of the missing square bracket[ ]_

## Question3-Answer


In [15]:
import pandas as pd

In [16]:
utility_info_df = pd.read_csv('https://bit.ly/HDSC-StageOneDataset')

In [17]:
utility_info_df

record_id  utility_id_ferc1  report_year  \
0        f1_fuel_1994_12_1_0_7                 1         1994   
1       f1_fuel_1994_12_1_0_10                 1         1994   
2        f1_fuel_1994_12_2_0_1                 2         1994   
3        f1_fuel_1994_12_2_0_7                 2         1994   
4       f1_fuel_1994_12_2_0_10                 2         1994   
...                        ...               ...          ...   
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                  rockport                coal       ton        5377489.0   
1      rockport total plant                coal       ton       10486945.0   
2                    gorgas                coal       ton        2978683.0   
3                     barry                coal       ton        3739484.0   
4                 chickasaw                 gas       mcf          40533.0   
...                     ...                 ...       ...              ...   
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                   16.590                      18.59   
1                   16.592                      18.58   
2                   24.130                      39.72   
3                   23.950                      47.21   
4                    1.000                       2.77   
...                    ...                        ...   
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                             18.53                1.121  
1                             18.53                1.120  
2                             38.12                1.650  
3                             45.99                1.970  
4                              2.77                2.570  
...                             ...                  ...  
29518                          4.78                9.030  
29519                          3.65                6.950  
29520                          4.77                8.990  
29521                         14.76                1.110  
29522                          0.00               11.680  

[29523 rows x 11 columns]

In [19]:
desc_data = utility_info_df.groupby(['fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].describe()

In [20]:
utility_info_df.columns

Index(['record_id', 'utility_id_ferc1', 'report_year', 'plant_name_ferc1',
       'fuel_type_code_pudl', 'fuel_unit', 'fuel_qty_burned',
       'fuel_mmbtu_per_unit', 'fuel_cost_per_unit_burned',
       'fuel_cost_per_unit_delivered', 'fuel_cost_per_mmbtu'],
      dtype='object')

In [21]:
desc_data

count         mean           std     min      25%  \
fuel_type_code_pudl                                                        
coal                  8547.0    67.421830    943.771433    0.00  23.9585   
gas                  11486.0    13.659397    163.286631    0.00   3.2300   
nuclear                818.0  4955.157002  14291.929749    0.00   1.6015   
oil                   8064.0   168.877086   2406.101479    0.00  37.6100   
other                  167.0    18.253856     29.016882 -119.04   2.6325   
waste                  441.0    19.518122     29.988017 -276.08  10.3350   

                         50%       75%         max  
fuel_type_code_pudl                                 
coal                 33.3180  44.66800   44033.320  
gas                   4.6500   6.96000   10490.000  
nuclear              32.5200  50.85525   69785.215  
oil                  68.9455  98.68925  139358.000  
other                 9.0640  20.05500     134.547  
waste                21.9330  32.65000     132.291

In [22]:
desc_data['mean'].min()

13.659396569737135

**Therefore the fuel type code that has lowest mean value cost per unit burned is gas**

## Question4-Answer

In [24]:
desc_data2 = utility_info_df['fuel_mmbtu_per_unit'].describe()

In [25]:
desc_data2

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

In [26]:
desc_data2['std'].round(2)

10.6

In [27]:
desc_data2['75%'].round(2)

17.01

**Therefore the standard deviation and 75% percentile of fuel_mmbtu_per_unit in two decimal place is 10.60 and 17.01**

## Question5-Answer

In [29]:
utility_info_df['fuel_qty_burned'].skew().round(2)

15.85

In [30]:
utility_info_df['fuel_qty_burned'].kurtosis().round(2)

651.37

**Therefore the skewness and kurtosis for the 'fuel_qty_burned' is 15.85 and 651.37 respectively**

## Question6-Answer

In [31]:
utility_info_df.isna().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [33]:
Total = utility_info_df['fuel_unit'].isna().sum()
Total

180

In [34]:
total_rc = utility_info_df['fuel_unit'].count()
total_rc

29343

In [35]:
Percentage = ((Total / total_rc) * 100).round(3)
Percentage

0.613

**Therefore: Feature:fuel_unit, Total:180, Percentage:0.613**

## Question7-Answer

**The column with missing value is a Categorical data and can use mode imputation techniques for missing values**

## Question8-Answer

In [36]:
Corr = utility_info_df.corr()
Corr

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [37]:
Corr['fuel_cost_per_unit_burned']

utility_id_ferc1               -0.037863
report_year                     0.013599
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
Name: fuel_cost_per_unit_burned, dtype: float64

In [38]:
Corr['fuel_cost_per_unit_burned'].min()

-0.03786270759997195

In [39]:
# assigning the values correlated value of fuel_cost_per_unit_burned with 
list1 = [i for i in Corr['fuel_cost_per_unit_burned'] ]
list1.sort()

In [40]:
list1[1:3]

[-0.018535416794825232, -0.010033808428852969]

**Therefore the features with the second and third lowest correlation to the fuel cost per unit burned are:
fuel_qty_burned and fuel_mmbtu_per_unit **

## Question9-Answer

In [41]:
utility_info_df.groupby(['report_year', 'fuel_type_code_pudl'])['fuel_cost_per_unit_burned'].sum().head(30)

report_year  fuel_type_code_pudl
1994         coal                    14984.572
             gas                     10792.496
             nuclear                227983.354
             oil                      9362.194
             other                     116.737
             waste                      52.762
1995         coal                    14571.785
             gas                      9087.905
             nuclear                384152.639
             oil                     17858.281
             other                     183.577
             waste                    -192.699
1996         coal                    12694.803
             gas                      1467.390
             nuclear                 66657.887
             oil                      5357.028
             other                     160.075
             waste                      88.092
1997         coal                    43742.178
             gas                      1518.471
             nuclear       

In [42]:
Percentage_change = ((14984.572 - 11902.597)/14984.572) * 100
Percentage_change

20.567654518260518

**Therefore the percentage change in fuel cost for the fuel type coal 
per unit burned in 1998 compared to 1994 is approximately 21 percent**

## Question10-Answer

In [43]:
grouped_data = utility_info_df.groupby(['report_year'])['fuel_cost_per_unit_delivered'].mean()
grouped_data

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

In [44]:
grouped_data.max()

11140.19723948813

**Therefore, 1997 is the year with the highest average fuel cost per unit delivered**